lib/xdrgen/generators/go.rb

```diff
@@ -461,7 +470,17 @@ def render_encode_to_body(out, var, type, self_encode:)
               out.puts "  if #{var} != nil {"
               var = "(*#{var})"
             end
-            var = "#{name type}(#{var})" if self_encode
+            if self_encode
+              newvar = "#{name type}(#{var})"
+              if type.resolved_type.is_a?(AST::Definitions::Typedef)
+                declared_type = type.resolved_type.declaration.type
+                # Fixed opaque types implement EncodeTo by pointer
+                if declared_type.is_a?(AST::Typespecs::Opaque) and declared_type.fixed?
+                  newvar = "(*#{name type})(&#{var})"
+                end
+              end
+              var = newvar
+            end
             out.puts "  err = #{var}.EncodeTo(e)"
```

Comment:
🤔 I don't remember why we even need the `type(var)`, the `var` should already be of type `type`. I might have introduced this unnecessarily, and if so, I think you could remove the self_encode line all together.

Reply:
Oh, I think I remember. This is for pointer types, such as `type PtrType *Type`, that need to cast to the underlying type before calling the function. It's sort of a shame that pointer types add this additional complexity. Maybe what I should have done is only add the type name wrapping in the pointer type case. This looks fine though and works, it's just adding more complexity to the logic you're adding when it probably doesn't need it.

Reply:
After viewing the diff you posted at https://github.com/stellar/xdrgen/pull/68#issuecomment-965789341 I realize this code is necessary for other types too. 👏🏻